developer-hub/src/reference_docs/defensive_programming.rs

```diff
@@ -0,0 +1,558 @@
+	#[docify::export]
+	#[test]
+	fn fixed_u64_block_computation_example() {
+		// Cores available per block
+		let supply = 10u128;
+		// Cores being ordered per block
+		let demand = 5u128;
+		// Calculate a very rudimentry on-chain price from supply / demand
+		let price = FixedU64::from_rational(demand, supply);
+
+		// 0.5 DOT per core
+		assert_eq!(price, FixedU64::from_float(0.5));
+
+		// Now, the story has changed - lots of demand means we buy as many cores as there
+		// available.  This also means that price goes up! For the sake of simplicity, we don't care
+		// about who gets a core - just about our very simple price model
+
+		// Cores available per block
+		let supply = 10u128;
+		// Cores being ordered per block
+		let demand = 19u128;
+		// Calculate a very rudimentary on-chain price from supply / demand
+		let price = FixedU64::from_rational(demand, supply);
+
+		// 1.9 DOT per core
+		assert_eq!(price, FixedU64::from_float(1.9));
+	}
+
+	#[docify::export]
+	#[test]
+	fn fixed_u64() {
+		// The difference between this and perthings is perthings operates within the relam of [0,
+		// 1] In cases where we need > 1, we can used fixed types such as FixedU64
+
+		let rational_1 = FixedU64::from_rational(10, 5); //" 200%" aka 2.
+		let rational_2 =
+			FixedU64::from_rational_with_rounding(5, 10, sp_arithmetic::Rounding::Down); // "50%" aka 0.50...
+
+		assert_eq!(rational_1, (2u64).into());
+		assert_eq!(rational_2.into_perbill(), Perbill::from_float(0.5));
+	}
+
+	#[docify::export]
+	#[test]
+	fn fixed_u64_operation_example() {
+		let rational_1 = FixedU64::from_rational(10, 5); // "200%" aka 2.
+		let rational_2 = FixedU64::from_rational(8, 5); // "160%" aka 1.6.
+
+		let addition = rational_1 + rational_2;
+		let multiplication = rational_1 * rational_2;
+		let division = rational_1 / rational_2;
+		let subtraction = rational_1 - rational_2;
+
+		assert_eq!(addition, FixedU64::from_float(3.6));
+		assert_eq!(multiplication, FixedU64::from_float(3.2));
+		assert_eq!(division, FixedU64::from_float(1.25));
+		assert_eq!(subtraction, FixedU64::from_float(0.4));
+	}
+	#[docify::export]
+	#[test]
+	fn bad_unwrap() {
+		let some_result: Result<u32, &str> = Ok(10);
+		assert_eq!(some_result.unwrap(), 10);
+	}
+
+	#[docify::export]
+	#[test]
+	fn good_unwrap() {
+		let some_result: Result<u32, &str> = Err("Error");
+		assert_eq!(some_result.unwrap_or_default(), 0);
+		assert_eq!(some_result.unwrap_or(10), 10);
+	}
+
+	#[docify::export]
+	#[test]
+	fn bad_collection_retrieval() {
+		let my_list = vec![1, 2, 3, 4, 5];
+		// THIS PANICS!
+		// Indexing on heap allocated values, i.e., vec, can be unsafe!
+		assert_eq!(my_list[5], 6)
+	}
+
+	#[docify::export]
+	#[test]
+	fn good_collection_retrieval() {
+		let my_list = vec![1, 2, 3, 4, 5];
+		// Rust includes `.get`, returning Option<T> - so lets use that:
+		assert_eq!(my_list.get(5), None)
+	}
+	#[docify::export]
+	#[test]
+	#[should_panic(expected = "Defensive failure has been triggered!")]
```

Comment:
```suggestion
	#[cfg_attr(debug_assertions, should_panic(expected = "Defensive failure has been triggered!")]
```
As you said above: it only panics with debug assertions and would fail on `cargo test --release`.